A few updates to free transaction creation and propagation

ptschip opened the pull request.

ptschip committed:
Allow a free transaction to be created up to the max txn size allowed

On the Bitcoin Cash network there are businesses that would like
to consolidate their coins on occasion but do not need to do so
in a hurry.  Sending free transactions is a way to make that happen
without costing anything.  And it also helps to reduce the size
of the UTXO.

ptschip committed:
Update limitfreerelay to be calculated by the default block size

As we move to larger blocks with Bitcoin Cash we need to have default
settings that change along with block size.  This prevents us from
having to do more code maintenance in the future.

ptschip committed:
Deprecate filtering by relaypriority for free transasctions.

This feature no longer exists in other implementataions and is
completely arbitrary in how the priority is selected to filter on.
In addition we already have filtering which chokes off free
transactions when the mempool fills up.  Therefore, this feature
is redundant and may end up filtering out free transactions which
are perfectly valid and could be mined within a few blocks.

ptschip committed:
Add missing policy.h include

sickpig commented on src/main.h:
```diff
@@ -133,8 +134,13 @@ static const uint32_t MAX_UNCONNECTED_HEADERS = 144;
+/** The maximum number of free transactions (in KB) that can enter the mempool per minute.
+ *  For a 1MB block we allow 15KB of free transactions per 1 minute.
+ */
+static const uint32_t DEFAULT_LIMITFREERELAY = DEFAULT_BLOCK_MAX_SIZE * 0.000015;
```
make sense

sickpig commented on src/main.h:
```diff
@@ -133,8 +134,13 @@ static const uint32_t MAX_UNCONNECTED_HEADERS = 144;
+ */
+static const uint32_t DEFAULT_LIMITFREERELAY = DEFAULT_BLOCK_MAX_SIZE * 0.000015;
+/** Subject free transactions to priority checking when entering the mempool */
+static const bool DEFAULT_RELAYPRIORITY = false;
```
what's the reason we need change it?

ptschip replied:
Other implementations don't have this anymore and we already have two other levels of spam control.  Also the filtering by priority is totally arbitrary...the calculation for the priority cutoff is defined as : inline double AllowFreeThreshold() { return COIN * 144 / 250; }

It's just a number somebody picked...if you're priority is below the threshold then you're not getting into the memory pool and your txn isn't getting fowarded.  Seems totally arbitrary to me and goes against the BU and BitcoinCash philosophy of allowing more free transactions into the network.

sickpig replied:
Agreed on the fact that the current prio threshold is pretty arbitrary.
AFAIK Bitcoin ABC still have `DEFAULT_RELAYPRIORITY` set to `false` and rely on the same value for the prio thres.
Regardless I can see it as a valuable change.

sickpig commented on src/wallet/wallet.h:
```diff
@@ -57,8 +58,13 @@ static const bool DEFAULT_SPEND_ZEROCONF_CHANGE = true;
+//! Largest (in bytes) free transaction we're willing to creat
+#ifdef BITCOIN_CASH
+//! We can allow a maximum sized free transaction in the Bitcoin Cash Network.
+static const unsigned int MAX_FREE_TRANSACTION_CREATE_SIZE = MAX_STANDARD_TX_SIZE;
```
first change you made 
```
static const uint32_t DEFAULT_LIMITFREERELAY = DEFAULT_BLOCK_MAX_SIZE * 0.000015;
```
we limited the amount of zero fee / hiprio txns to be relayed to 15KB per minute, hence it would take ~66 minutes to get a 1MB transactions to be relayed between two nodes without any other zero fee txns passing by during that period. Am I correct?

ptschip replied:
@sickpig not sure what you mean about ~66 mins?.  The only change is that the limitfreerelay value will rise with blocksize, but there is no change in current behavior at the 1MB block level.  Basically there is a 10minute window which decays..so every 10 minutes you can send/route 150KB of free txns in that window. (or 15KB per minute), but whether they get mined or not is up to the miners.

sickpig replied:
`MAX_FREE_TRANSACTION_CREATE_SIZE = 1MB`
implies that I could create a free transaction as big as 1MB, but as you said the net (by default) can only send/route 150KB of free every 10 minutes. 

I didn't looked at the code but suppose that the "counter" that keep track of free txns is not restarted every new  block.

That means that I can create a 1MB free transaction and that it would be succesfully relayed to another node only if no other free txns is sent/routed for at 6/7 blocks (1MB / 15 KB/min) = 66 minutes.

Am I wrong?

ptschip replied:
@sickpig The biggest free txn is only MAX_STANDARD_TX_SIZE or 100KB...i think you didn't see that and thought it was max block size.

sickpig replied:
I mean shouldn't we anchor `MAX_FREE_TRANSACTION_CREATE_SIZE` to `DEFAULT_BLOCK_MAX_SIZE` like we did for `DEFAULT_LIMITFREERELAY`  ?

ptschip replied:
@sickpig also it' a little confusing because the decay window is actually every 10 minutes, not every minute...in the past they should have really made limit free relay 150  instead of 15 (and they they multiply by 10 at some point), but I hesitate to tidy that up because it involves a little refactoring.   So in actual fact it's 150KB per 10 minutes.

ptschip replied:
@sickpig oh i see, about the anchoring...no i don't think so ...the max txn is 100KB and is a routing policy across the network...we can't just arbitrarily up that figure otherwise our txns won't get routed. And really there is very little use case for txns greater than 100KB.  I think we had this discussion during the SigOps round of discussions.

sickpig replied:
ouch you are right, I read 1MB when it was 100KB. Sorry for the noise.

sickpig commented:
It seems that we have a problem with `abandonconflict.py`  see:
https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/316641484#L2373

are you able to reproduce it locally?

ptschip commented:
@sickpig yup, thanks...I have to sort through that.  There's a bug in there somewhere, they seem to depend on relaypriority to make that testcase work, although it shouldn't be necessary.

ptschip committed:
Properly initialize variables

ptschip committed:
abandonconflict.py still relies on -relaypriority

 Use limitfreerelay to prevent txns from getting into the mempool.

 Also added function LimitFree() which determines what the node
 policy is regarding free transactions and is used to pass fLimitFree
 to AcceptToMemoryPoolWorker.  If -limitfreerelay is greater than zero
 then during startup or when a re-org happens we allow all free
 transactions into the memory pool because we can not know when those
 free transactions were created.  If however, the node policy is
 set such that no free transactions are allowed then fLimitFree is set
 to true and they are then subject to rate limiting.

ptschip committed:
Only check priority if -relaypriority is true

When sending free transactions , only check the priority is
enough if -relaypriority is true.

ptschip committed:
Do not allow free trasactions if -limitfreerely is zero

If this node has set a node policy for limitfreerelay of zero
then there is no point in allowing the wallet to create and send
free transactions because they will not be accepted to the mempool
nor routed by this node.

sickpig commented on src/wallet/wallet.cpp:
```diff
@@ -2329,6 +2329,13 @@ bool CWallet::CreateTransaction(const vector<CRecipient>& vecSend, CWalletTx& wt
+                    strFailReason =
+                        _("You can not send free transactions if you have configured a -limitfreerelay of zero");
+                    return false;
+                }
```
What if the zero fee txns is created using "old" utxo so that the resulting txns could be considered an high priority one?

ptschip replied:
This is just handling a rare edge case.  If you have configured limitfreerelay of zero..that means you're not going to nor intend to route "any" free transactions...so if you created one with your wallet , it won't get routed by your very own node and this is just warning that you'll need to adjust your limitfreerelay if you want to send free transactions.  Priority doesn't matter in this context.   But now that you mention it, I should break that out into it's own "if" statement rather than an else if...i'll change that.

gandrewstone replied:
this is great.  I think from a usability perspective, we should not allow a tx to be created that our own node would normally reject.  This will stop people from accidentally making zombie tx.

ptschip committed:
Use separate if statement

When checking for whether free transactions can be routed by
our node, before actually going ahead a creating one, we should
do this in a separate if statement othewise we could miss this
check when -relaypriority is in effect.

gandrewstone commented on src/wallet/wallet.cpp:
```diff
@@ -1333,7 +1333,7 @@ void CWallet::ReacceptWalletTransactions()
         CWalletTx& wtx = *(item.second);
 
-        wtx.AcceptToMemoryPool(false);
+        wtx.AcceptToMemoryPool(LimitFree());
```
I thought that the idea was to accept your own wallet transactions regardless of your relay policy.  

I can see how it may be a good idea to not accept or at least warn if a new transaction breaks your own relay rules, but in this case we'd be blocking an old transaction.  I don't think that this makes sense.  If we won't put it in our own mempool it'll never be relayed and therefore end up as a zombie tx forever.  It might be hard from a UX perspective to clean them up here though -- it'll look like tx just magically disappear (and they might still exist in the network).  I think that its best to accept all wallet tx into our own mempool and then have the user clean up ones that will never confirm separately.

ptschip replied:
yes, good point.  If the user had sent a lot of free transactions and then restarted their node...then they would be subject here to rate limiting and may or may not get accepted to the mempool.  if for instance they had sent over the course of many hours enough free transactions > 150KB then some of them would not get into the mempool.  So yes you're right we should accept these into the pool regardless of how many there are.

gandrewstone commented on src/main.cpp:
```diff
@@ -1166,6 +1166,15 @@ bool IsDAAEnabled(const CChainParams &chainparams, const CBlockIndex *pindexPrev
 }
 #endif
 
+bool LimitFree()
```
"LimitFree" implies that it'll return the limit.  I'd prefer this to be called something like "AreFreeTxDisallowed".

ptschip replied:
Good idea, but it actually it doesn't mean that exactly either...it really means, are free transactions subject to rate limiting, so maybe something like :  "AreFreeTxnsRateLimited", will do the trick.  I'll make the fix.

ptschip replied:
hmmm, actually it means both:  if it returns true then it means free txns are disallowed, if it returns false then it means not subject to rate limiting...so I think your AreFreeTxnsDisallowed, is more clear and closer to the truth.

ptschip committed:
Do not subject txns to -limitfreerelay when reaccepting wallet txns

ptschip committed:
Change LimtFree() to AreFreeTxnsDisallowed()

This clarifies the intent of this function.

ptschip commented:
@gandrewstone I addressed your comments and updated the PR.

ptschip committed:
Use relaypriority in abandonconflict.py

In order for this test to work we have to prevent transactions
from entering the mempool on startup. Currently the only way
to do this is to turn on relaypriority.  In the future when
relaypriority is removed from the software we'll have to find
another way to make this test work.

ptschip committed:
fix formatting
